lib/beautiful/colorful.lua

```diff
@@ -0,0 +1,1395 @@
+        css.comma,
+        perc(capt(css.integer)), -- lightness
+        css.comma,
+        perc(capt(css.integer)), -- alpha
+        css.eol
+        }, "")
+    },
+    -- }}}
+    -- {{{ hsla(d % % / f) integers 0-360, percentage 0-100, float 0-1
+    { "hslxf", table.concat({
+        css.hsla,
+        capt(css.integer), -- hue
+        css.space,
+        perc(capt(css.integer)), -- saturation
+        css.space,
+        perc(capt(css.integer)), -- lightness
+        css.slash,
+        capt(css.float), -- alpha
+        css.eol
+        }, "")
+    },
+    -- }}}
+    -- {{{ hsla(d % % / %) integers 0-360, percentage 0-100
+    { "hslxp", table.concat({
+        css.hsla,
+        capt(css.integer), -- hue
+        css.space,
+        perc(capt(css.integer)), -- saturation
+        css.space,
+        perc(capt(css.integer)), -- lightness
+        css.slash,
+        perc(capt(css.integer)), -- alpha
+        css.eol
+        }, "")
+    },
+    -- }}}
+    -- }}}
+}
+-- }}} Color patterns
+
+--- Create a new color instance from a string
+-- @tparam string str A CSS color compatible string.
+-- @return A new color object, or nil if invalid
+-- @usage local c = colorful.color.from.string("rgb(255, 204, 153)")
+function color.from.string(str)
+    for _,pat in pairs(color_patterns) do
+        local t = pat[1]
+        local m = {str:match(pat[2])}
+        if m[1] ~= nil then
+            local th = t:sub(1,3)
+            local tv = t:sub(4,4)
+            local ta = #t == 5 and t:sub(5,5) or nil
+            if th == "hex" then
+                -- r,g,b convert to number
+                local r = tonumber(m[1], 16)
+                local g = tonumber(m[2], 16)
+                local b = tonumber(m[3], 16)
+                local a = 0xFF
+                -- if alpha present
+                if ta ~= nil then
+                    a = tonumber(m[4], 16)
+                end
+                -- if type is shorthand
+                if tv == "s" then
+                    r = r * 16 + r
+                    g = g * 16 + g
+                    b = b * 16 + b
+                    a = a * 16 + a
+                end
+                return color.from.rgba(r,g,b,a)
+            elseif th == "rgb" then
+                local r = tonumber(m[1])
+                local g = tonumber(m[2])
+                local b = tonumber(m[3])
+                local a = 255
+                -- if values percentages
+                if tv == "p" then
+                    r,g,b = r/100 * 255, g/100 * 255, b/100 * 255
+                end
+                -- check for alpha
+                if ta ~= nil then
+                    a = tonumber(m[4])
+                    if ta == "p" then
+                        a = a/100
+                    end
+                    a = a * 255 -- both float and percentage need this
+                end
+                if r == nil or g == nil or b == nil or a == nil then return nil end
```

Comment:
Why this line? The `hex` case does not have it.

Reply:
You're right! Mistake on my part, I can add the check to the `hex` section.